File: scripts/migrate_subscription_status.py
"""
Script de migración one-shot: normaliza status de suscripciones a minúsculas.

Uso:
  # Dry-run (solo muestra qué haría):
  python scripts/migrate_subscription_status.py

  # Ejecutar migración real:
  python scripts/migrate_subscription_status.py --apply

Requiere MONGODB_URL como variable de entorno o usa la de settings.
"""
import sys
import os

# Agregar el path del backend para importar settings
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'backend'))

from pymongo import MongoClient

# Intentar importar settings, o usar env var directamente
try:
    from app.config.settings import settings
    MONGODB_URL = settings.MONGODB_URL
    MONGODB_DATABASE = settings.MONGODB_DATABASE
except Exception:
    MONGODB_URL = os.environ.get("MONGODB_URL", "mongodb://localhost:27017")
    MONGODB_DATABASE = os.environ.get("MONGODB_DATABASE", "cuenly")

COLLECTION = "user_subscriptions"

STATUS_MAP = {
    "ACTIVE": "active",
    "PAST_DUE": "past_due",
    "CANCELLED": "cancelled",
    "EXPIRED": "expired",
}


def migrate(apply: bool = False):
    client = MongoClient(MONGODB_URL, serverSelectionTimeoutMS=5000)
    db = client[MONGODB_DATABASE]
    coll = db[COLLECTION]

    print(f"Conectado a: {MONGODB_DATABASE}")
    print(f"Colección: {COLLECTION}")
    print(f"Modo: {'APLICAR CAMBIOS' if apply else 'DRY-RUN (solo lectura)'}")
    print("=" * 60)

    total_updated = 0

    for old_status, new_status in STATUS_MAP.items():
        count = coll.count_documents({"status": old_status})
        if count > 0:
            print(f"  '{old_status}' → '{new_status}': {count} documentos")
            if apply:
                result = coll.update_many(
                    {"status": old_status},
                    {"$set": {"status": new_status}}
                )
                print(f"    → Actualizados: {result.modified_count}")
                total_updated += result.modified_count
        else:
            print(f"  '{old_status}' → '{new_status}': 0 documentos (OK)")

    print("=" * 60)

    if apply:
        print(f"Total actualizados: {total_updated}")
    else:
        print("Dry-run completado. Ejecuta con --apply para aplicar cambios.")

    # Resumen final de status en BD
    print("\nDistribución actual de status:")
    pipeline = [{"$group": {"_id": "$status", "count": {"$sum": 1}}}]
    for doc in coll.aggregate(pipeline):
        print(f"  {doc['_id']}: {doc['count']}")

    client.close()


if __name__ == "__main__":
    do_apply = "--apply" in sys.argv
    migrate(apply=do_apply)

File: scripts/migrate_subscription_status.js
// Migration: Normalizar status de suscripciones a minúsculas
// Ejecutar con: mongosh <connection_string> scripts/migrate_subscription_status.js
// O en Docker: docker compose exec mongodb mongosh cuenly scripts/migrate_subscription_status.js

print("=== Migración: Normalizar subscription status a minúsculas ===");

// Verificar estado actual
print("\nEstado ANTES de la migración:");
print("  ACTIVE:", db.user_subscriptions.countDocuments({ status: "ACTIVE" }));
print("  active:", db.user_subscriptions.countDocuments({ status: "active" }));
print("  PAST_DUE:", db.user_subscriptions.countDocuments({ status: "PAST_DUE" }));
print("  past_due:", db.user_subscriptions.countDocuments({ status: "past_due" }));
print("  CANCELLED:", db.user_subscriptions.countDocuments({ status: "CANCELLED" }));
print("  cancelled:", db.user_subscriptions.countDocuments({ status: "cancelled" }));
print("  EXPIRED:", db.user_subscriptions.countDocuments({ status: "EXPIRED" }));
print("  expired:", db.user_subscriptions.countDocuments({ status: "expired" }));

// Migrar
var r1 = db.user_subscriptions.updateMany({ status: "ACTIVE" }, { $set: { status: "active" } });
print("\nACTIVE -> active: " + r1.modifiedCount + " documentos actualizados");

var r2 = db.user_subscriptions.updateMany({ status: "PAST_DUE" }, { $set: { status: "past_due" } });
print("PAST_DUE -> past_due: " + r2.modifiedCount + " documentos actualizados");

var r3 = db.user_subscriptions.updateMany({ status: "CANCELLED" }, { $set: { status: "cancelled" } });
print("CANCELLED -> cancelled: " + r3.modifiedCount + " documentos actualizados");

var r4 = db.user_subscriptions.updateMany({ status: "EXPIRED" }, { $set: { status: "expired" } });
print("EXPIRED -> expired: " + r4.modifiedCount + " documentos actualizados");

// Verificar estado final
print("\nEstado DESPUÉS de la migración:");
print("  active:", db.user_subscriptions.countDocuments({ status: "active" }));
print("  past_due:", db.user_subscriptions.countDocuments({ status: "past_due" }));
print("  cancelled:", db.user_subscriptions.countDocuments({ status: "cancelled" }));
print("  expired:", db.user_subscriptions.countDocuments({ status: "expired" }));

var total = r1.modifiedCount + r2.modifiedCount + r3.modifiedCount + r4.modifiedCount;
print("\n=== Total: " + total + " documentos migrados ===");
